### PR TITLE
NP-46888 Update who can delete messages

### PR DESCRIPTION
**Pull request by joachimjorgensen**
[NP-46888](https://unit.atlassian.net/browse/NP-46888)

Up until now, only `messageOwner` could delete messages. The problem with this, is that `messageOwner` is the same as the `ticketOwner`, which in turn is the `publicationOwner`. Now, it isn't necessarily the `publicationOwner` that creates a message. It could be a contributor or even a curator. Therefore, the `messageSender` is now allowed to delete a message.

Moreover, a curator with correct `accessRights` are also allowed to delete messages within their institution, to help clean up if necessary. 

[NP-46888]: https://unit.atlassian.net/browse/NP-46888?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: publication-commons/src/main/java/no/unit/nva/publication/service/impl/MessageService.java [hunk 1]
```diff
@@ -85,6 +97,28 @@ public void deleteMessage(UserInstance userInstance, Message message) throws Una
...
+    private boolean canManageMessage(Message message, UserInstance userInstance) throws NotFoundException {
+        if (userInstance.isSender(message)) {
+            return true;
+        }
+
+        return isCuratorForMessage(message, userInstance);
+    }
+
+    private boolean isCuratorForMessage(Message message, UserInstance userInstance) throws NotFoundException {
+        if (!message.getCustomerId().equals(userInstance.getCustomerId())) {
```
Comment (brinxmat) [inline publication-commons/src/main/java/no/unit/nva/publication/service/impl/MessageService.java:109]: This implies that we ignore fine-grained restrictions based on supposed hierarchies within an organisation, which for me is fine

**Review by brinxmat — APPROVED** [review]

File: publication-commons/src/test/java/no/unit/nva/publication/service/impl/MessageServiceTest.java [hunk 2]
```diff
@@ -117,10 +138,59 @@ void shouldThrowUnauthorizedWhenAttemptingToDeleteMessageUserDoesNotOwn()
...
+            assertThrows(UnauthorizedException.class, () -> messageService.deleteMessage(publishingCurator, persistedMessage));
+        }
+
+        if (ticketType == PublishingRequestCase.class) {
+            assertDoesNotThrow(() -> messageService.deleteMessage(publishingCurator, persistedMessage));
+            assertThrows(UnauthorizedException.class, () -> messageService.deleteMessage(publishingCuratorFromAnotherInstitution, persistedMessage));
+            assertThrows(UnauthorizedException.class, () -> messageService.deleteMessage(doiCurator, persistedMessage));
+            assertThrows(UnauthorizedException.class, () -> messageService.deleteMessage(supportCurator, persistedMessage));
+        }
+    }
```
Comment (truhacevkir) [inline publication-commons/src/test/java/no/unit/nva/publication/service/impl/MessageServiceTest.java:180]: I would refactor this test to two tests, the one which allows deleting, and another one which not.

You can use following method source for test which tests allowance:

```
    public static Stream<Arguments> ticketTypeAccessRightsProvider() {
        return Stream.of(Arguments.of(DoiRequest.class, MANAGE_DOI),
                         Arguments.of(PublishingRequestCase.class, MANAGE_PUBLISHING_REQUESTS),
                         Arguments.of(UnpublishRequest.class, MANAGE_PUBLISHING_REQUESTS),
                         Arguments.of( GeneralSupportRequest.class, SUPPORT));
    }
```
Your test will look like this:
```
  @ParameterizedTest
    @MethodSource("ticketTypeAccessRightsProvider")
    void shouldAllowCuratorToDeleteMessageWhenTicketHasCorrectType(Class<? extends TicketEntry> ticketType,
                                                                   AccessRight accessRight) throws ApiGatewayException {
        var publication = TicketTestUtils.createPersistedPublicationWithOwner(
            PublicationStatus.PUBLISHED, owner, resourceService);
        var ticket = TicketTestUtils.createPersistedTicket(publication, ticketType, ticketService);
        var persistedMessage = messageService.createMessage(ticket, owner, randomString());
        var curator = randomUserInstance(accessRight, owner.getCustomerId());

        assertDoesNotThrow(() -> messageService.deleteMessage(curator, persistedMessage));
    }
```

File: publication-commons/src/main/java/no/unit/nva/publication/service/impl/MessageService.java (hunk 1 above)
Comment (joachimjorgensen) [reply publication-commons/src/main/java/no/unit/nva/publication/service/impl/MessageService.java:109]: Yes, that is intentional 👍

File: publication-commons/src/test/java/no/unit/nva/publication/service/impl/MessageServiceTest.java (hunk 2 above)
Comment (truhacevkir) [reply publication-commons/src/test/java/no/unit/nva/publication/service/impl/MessageServiceTest.java:180]: You need rule for unpublishRequest to this method to make test pass:
`private boolean isCuratorOnTicket(Message message, UserInstance userInstance)` 

I do not know the requirements here:
`case UnpublishRequest unpublishRequest-> true;
`

Comment (truhacevkir) [reply publication-commons/src/test/java/no/unit/nva/publication/service/impl/MessageServiceTest.java:179]: Just an example :)

File: publication-commons/src/main/java/no/unit/nva/publication/service/impl/MessageService.java [hunk 3]
```diff
@@ -85,6 +98,29 @@ public void deleteMessage(UserInstance userInstance, Message message) throws Una
...
+            return false;
+        }
+
+        var ticket = ticketService.fetchTicketByIdentifier(message.getTicketIdentifier());
+        return switch (ticket) {
+            case PublishingRequestCase publishingRequest -> userInstance.getAccessRights().contains(AccessRight.MANAGE_PUBLISHING_REQUESTS);
+            case GeneralSupportRequest supportRequest -> userInstance.getAccessRights().contains(AccessRight.SUPPORT);
+            case DoiRequest doiRequest -> userInstance.getAccessRights().contains(AccessRight.MANAGE_DOI);
+            case UnpublishRequest unpublishRequest -> userInstance.getAccessRights().contains(AccessRight.MANAGE_RESOURCES_STANDARD);
```
Comment (joachimjorgensen) [inline publication-commons/src/main/java/no/unit/nva/publication/service/impl/MessageService.java:119]: Using the same accessRight as CuratorPermissionStrategy

**Review by truhacevkir — APPROVED** [review]

**Pull request merged**